pkg/ring/basic_lifecycler.go

```diff
@@ -0,0 +1,474 @@
+package ring
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/cortexproject/cortex/pkg/ring/kv"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+type BasicLifecyclerDelegate interface {
+	// OnRingInstanceRegister is called while the lifecycler is registering the
+	// instance within the ring and should return the state and set of tokens to
+	// use for the instance itself.
+	OnRingInstanceRegister(lifecycler *BasicLifecycler, ringDesc Desc, instanceExists bool, instanceID string, instanceDesc IngesterDesc) (IngesterState, Tokens)
+
+	// OnRingInstanceTokens is called once the instance tokens are set and are
+	// stable within the ring (honoring the observe period, if set).
+	OnRingInstanceTokens(lifecycler *BasicLifecycler, tokens Tokens)
+
+	// OnRingInstanceStopping is called while the lifecycler is stopping. The lifecycler
+	// will continue to hearbeat the ring the this function is executing and will proceed
+	// to unregister the instance from the ring only after this function has returned.
+	OnRingInstanceStopping(lifecycler *BasicLifecycler)
+}
+
+type OnRingInstanceRegister func(lifecycler *BasicLifecycler, ringDesc Desc, instanceExists bool, instanceID string, instanceDesc IngesterDesc) (IngesterState, Tokens)
+type OnRingInstanceTokensChanged func(lifecycler *BasicLifecycler, tokens Tokens)
+type OnRingInstanceStopping func(lifecycler *BasicLifecycler)
+
+type BasicLifecyclerConfig struct {
+	InstanceID          string
+	InstanceAddr        string
+	InstancePort        int
+	InstanceZone        string
+	KVStore             kv.Config
```

Comment:
Do you think KVStore can be removed from lifecycler config? I think Lifecycler should be passed a KV store, and not create one on its own, but maybe it would complicate planned usage. (Right now, only `NewBasicLifecyclerWithStoreClient` is used though)

Reply:
Creating a KVStore client is very easy (few lines of code) and I agree we should always pick it in input to make code that uses `BasicLifecycler` easier to test.